### PR TITLE
Implement Alpha Mods

### PR DESCRIPTION
**Pull request by MegaPiggy**
https://github.com/Raicuparta/ow-mod-db/pull/299

### CODE REVIEW

File: app/components/Mods/AlphaSwitch.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { Switch, Tooltip, FormLabel } from '@material-ui/core';
+import { useSetRecoilState, useRecoilValue } from 'recoil';
+import { alphaState } from '../../store';
+
+const AlphaSwitch: React.FunctionComponent = () => {
```
Comment (Raicuparta) [inline app/components/Mods/AlphaSwitch.tsx:6]: So, sorry for changing my mind, but I think this switch shouldn't go here after all. A tiny, tiny fraction of people using mods care about the alpha (from what I can tell in the website stats), so I don't think the button should be so prominent.

I'd prefer if this switch is moved to the settings tab. It can replace the "Show Alpha Settings" switch. So you can go to the settings tab, enable "View Mods for Outer Wilds Alpha", which simultanesly expands the alpha settings and switches to alpha mode.

**Comment by Raicuparta** [comment]
Looks like you're not using ESLint nor Prettier:
- Install the ESLint VSCode extension;
- Install the Prettier VSCode extension (maybe not needed since this project uses prettier through eslint);
- Configure VSCode to auto-format on save (optional but recommended);
- Do `yarn run lint-fix` to fix all the auto-fixable problems;
- Do `yarn run lint` to see a list of all the problems that you need to fix manually.

File: app/components/Mods/ModTableRow.tsx [hunk 2]
```diff
@@ -181,6 +187,14 @@ const ModTableRow: React.FunctionComponent<Props> = ({ mod }) => {
     if (isModConflicting) {
       return modsText.conflictingModWarning(conflictingMods.join(', '));
     }
+    if (isInstalled(mod) && wrongBepinExVersion){
+      if (!bepInEx || !isInstalled(bepInEx)) {
+        return modsText.missingDependencyWarning("bbepis.BepInEx");
```
Comment (Raicuparta) [inline app/components/Mods/ModTableRow.tsx:192]: the `bbepis.BepInEx` string is showing in a lot of places. Better move it to some separate file with uniqueName constants, together with `Locochoco.CMOWA` and `Alek.OWML` (this last one not your fault but might as well do it now)

File: app/components/Mods/AlphaSwitch.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    const setAlpha = useSetRecoilState(alphaState);
+    const alpha = useRecoilValue(alphaState);
+    return (
+        <div>
+            <Tooltip title={'View the mods created for Outer Wilds Alpha.'}>
```
Comment (Raicuparta) [inline app/components/Mods/AlphaSwitch.tsx:11]: I know I just said to remove this switch, but remember to have all translatable text in `static-text.ts`

File: app/components/Mods/ModsPage.tsx [hunk 4]
```diff
@@ -10,14 +10,21 @@ import {
...
+import { alphaState } from '../../store';
 
 const ModsPage: React.FunctionComponent = () => {
-  const enabledMods = useRecoilValue(enabledModList);
-  const installedMods = useRecoilValue(installedModList);
-  const notInstalledMods = useRecoilValue(notInstalledModList);
-  const requiredMods = useRecoilValue(requiredModList);
+  const alpha = useRecoilValue(alphaState);
+  
+  const enabledMods = useRecoilValue(alpha ? enabledAlphaModList : enabledModList);
```
Comment (Raicuparta) [inline app/components/Mods/ModsPage.tsx:24]: Instead of constantly checking the `alpha` flag every time you need to access these state values, you should instead move the flag check to inside the selector. This is basically what selectors are for.

In this case you'd just use `enabledModList` for both the alpha and non-alpha mods. And then inside the selector you check the `alphaState` and return the correct list accordingly. Do this with all of the lists that depend on whether alpha mode is enabled or not.

File: app/components/Mods/ModsPage.tsx [hunk 5]
```diff
@@ -36,11 +43,14 @@ const ModsPage: React.FunctionComponent = () => {
           highlighted
         />
         {isEmpty && (
-          <Card>
-            <Typography variant="h6" align="center">
-              {modsText.emptyModList}
-            </Typography>
-          </Card>
+          <div>
+            <br/>
```
Comment (Raicuparta) [inline app/components/Mods/ModsPage.tsx:47]: I didn't check exactly why you needed this, but doesn't look good. Don't use `br` to add spacing, `br` has a semantic meaning. Either add a top margin to the Card component, or add a Box wrapper with that margin.

File: app/components/Mods/ModsToolbar.tsx [hunk 6]
```diff
@@ -18,8 +19,9 @@ const useStyles = makeStyles((theme) => ({
 
 const ModsToolbar: React.FunctionComponent = () => {
   const styles = useStyles();
+  const alpha = useRecoilValue(alphaState);
```
Comment (Raicuparta) [inline app/components/Mods/ModsToolbar.tsx:22]: it's convention to prefix flags to make code more readable (`is`, `has`, etc). In this case it should be `isAlpha`. Also, don't be afraid of long variable names. The more descriptive the better. Like `isAlphaModsEnabled` or whatever you feel drescribes more precisely what this flag does, to for instance distinguish it from. Especially when you already have a `mod.isAlpha`, so you want to always be able to very clearly tell what you're looking at.

File: app/components/Settings/AlphaDownload.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      <Button
+        variant="contained"
+        onClick={handleClick}
+        startIcon={<DownloadIcon />}
+      >
+        {'Download Outer Wilds Alpha'}
+      </Button>
+    </ListItem>
+  );
+};
```
Comment (Raicuparta) [inline app/components/Settings/AlphaDownload.tsx:32]: Instead of a direct download button, let's just have a button that points to `https://outerwildsmods.com/outer-wilds-alpha/`. This way we don't have to change two places if we ever change how to download the alpha (unless we make it downloadable via the database like a mod), and we reduce the chance of getting into trouble. The button text should of course also change to reflect this.

File: app/components/Settings/SettingsPage.tsx [hunk 8]
```diff
@@ -106,6 +122,7 @@ const Settings = () => {
                 </React.Fragment>
               )
           )}
+          <AlphaDownload />
```
Comment (Raicuparta) [inline app/components/Settings/SettingsPage.tsx:125]: this should be inside the alpha settings expand

File: app/components/Settings/SettingsPage.tsx [hunk 9]
```diff
@@ -73,19 +75,33 @@ const settingsInputs: readonly SettingsInput[] = [
...
     isAdvanced: true,
   },
+  {
+    key: 'showAlphaSettings',
+    type: SettingType.Switch,
+  },
+  {
+    key: 'alphaPath',
+    type: SettingType.Path,
+    isAlpha: true,
```
Comment (Raicuparta) [inline app/components/Settings/SettingsPage.tsx:85]: Pressing "Find" and then cancelling in the popup for some reason makes "showa dvanced settings" toggle back to false. It also makes the "Save" button show up even though the input is empty (might be an old manager bug). Same thing happens with the OWML path input too now.

File: app/components/TopBar/TopBar.tsx [hunk 10]
```diff
@@ -1,9 +1,15 @@
...
+import { useRecoilValue } from 'recoil';
+
+import { alphaState } from '../../store';
 
 import StartGameButton from './StartGameButton';
+import StartAlphaButton from './StartAlphaButton';
 import AppTabs from './AppTabs';
+import { tabList } from './AppTabs';
 import LoadingSuspense from '../LoadingSuspense';
+import { globalText } from '../../helpers/static-text';
```
Comment (Raicuparta) [inline app/components/TopBar/TopBar.tsx:12]: you added some imports that you don't use. Would have been caught by either ESLint or just normal TypeScript warnings.

File: app/config.json [hunk 11]
```diff
@@ -10,7 +10,10 @@
     "showAdvancedSettings": false,
     "disableModWarnings": {},
     "modDatabaseUrl": "https://raw.githubusercontent.com/Raicuparta/ow-mod-db/master/database.json",
-    "alertSourceUrl": "https://raw.githubusercontent.com/Raicuparta/ow-mod-db/source/alert.json"
+    "alertSourceUrl": "https://raw.githubusercontent.com/Raicuparta/ow-mod-db/source/alert.json",
+    "showAlphaSettings": true,
```
Comment (Raicuparta) [inline app/config.json:14]: this should default to false

File: app/helpers/static-text.ts [hunk 12]
```diff
@@ -156,6 +173,20 @@ export const settingsText = {
     tooltip:
       'Prevents this mod from showing warnings in the Mod Manager when starting the game',
   },
+  showAlphaSettings: {
+    label: 'Show Alpha Settings',
+    tooltip: '',
+  },
+  alphaPath: {
+    label: 'Alpha game folder (contains alpha game exe)',
```
Comment (Raicuparta) [inline app/helpers/static-text.ts:181]: this text is a bit too big to fit comfortably in a min width window, maybe move the "contains alpha game exe" thing to the description

File: app/helpers/version-compare.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+export const versionComparer = {
+    areEqual: areEqual,
+    isGreaterThan: isGreaterThan,
+    isGreaterThanOrEqualTo: isGreaterThanOrEqualTo,
+    isLessThan: isLessThan,
+    isLessThanOrEqualTo: isLessThanOrEqualTo,
+    isBetweenMinAndMax : isBetweenMinAndMax
+};
```
Comment (Raicuparta) [inline app/helpers/version-compare.ts:92]: Oof, this whole thing is just for checking the BepInEx version, right? Seems a bit overkill when most likely the correct version will be downloaded (CAMOWA could probably just print a warning if an incompatible version is installed).

If you want to keep it anyway: use descriptive names, no single-letter variables (`versionA` and `versionB` instead of `a` and `b` or `v1` and `v2`).

Example:
```ts
function versionCompare(v1 : string, v2 : string) {
    var version1 = cleanID(v1);
```
becomes
```ts
function versionCompare(versionA: string, versionB: string) {
    var cleanVersionA = cleanID(versionA);
```

Comment (MegaPiggy) [reply app/helpers/version-compare.ts]: I'll just remove the BepInEx version thingy, it isn't needed anymore.

File: app/components/Settings/SettingsPage.tsx (hunk 9 above)
Comment (MegaPiggy) [reply app/components/Settings/SettingsPage.tsx:85]: It seems to reset everything back to default settings.

File: app/components/Mods/ModsPage.tsx (hunk 4 above)
Comment (MegaPiggy) [reply app/components/Mods/ModsPage.tsx]: I'm not able to add that to a selector. It breaks the entire mod manager.
image

File: app/hooks/use-mods-directory-watcher.ts [hunk 14]
```diff
@@ -15,11 +19,16 @@ export function useModsDirectoryWatcher(owmlPath: string, handler: Handler) {
 
     debugConsole.log('useEffect: useModsDirectoryWatcher');
     const path = `${owmlPath}/Mods`;
+    const patha = `${alphaPath}/BepInEx/plugins`;
```
Comment (Raicuparta) [inline app/hooks/use-mods-directory-watcher.ts:22]: First of all, bad naming all around this file. But you're repeating a lot of unnecessary code anyway. Just fetch the `isAlphaMode` value from the settings store (in fact, you can probably fetch owmlPath and alphaPath from the store too instead of passing them as arguments). Then do 
```ts
const path = isAlphaMode ? `${alphaPath}/BepInEx/plugins` : `${owmlPath}/Mods`
```
The rest of the file can be as it was before.

File: app/package.json [hunk 15]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "ow-mod-manager",
   "productName": "OuterWildsModManager",
-  "version": "1.10.1",
+  "version": "1.11.0",
```
Comment (Raicuparta) [inline app/package.json:4]: revert the version change, I keep version changes in dedicated commits. Makes it easier to avoid conflicts and to track version changes in the commit history.

File: app/services/get-local-mods.ts [hunk 16]
```diff
@@ -30,81 +30,214 @@ function getOwml(owmlPath: string) {
...
+      : undefined,
+    isEnabled: true,
+    isRequired: true,
+    errors: [],
+    dependencies: ['bbepis.BepInEx'],
+    addons: [],
+    isAlpha: true,
+  };
+  return cmowa;
+}
```
Comment (Raicuparta) [inline app/services/get-local-mods.ts:77]: The reason OWML had some of this strange hardcoded stuff is because early OWML versions had no manifest.json. If these required items always have their own manifests, then the hardcoded strings are probably not needed. I'm just making sure you didn't just copy this part without thinking because you saw it being done for OWML. Don't know if it's actually needed or not. I leave it up to you.

File: app/services/get-local-mods.ts [hunk 17]
```diff
@@ -30,81 +30,214 @@ function getOwml(owmlPath: string) {
...
-        modPath,
-        name: modDirectoryName,
-        uniqueName: modDirectoryName,
-        localVersion: '-',
-        addons: [],
-      });
-    }
-  });
+    });
+  }
```
Comment (Raicuparta) [inline app/services/get-local-mods.ts:240]: Everything inside the `if (alphaPath)` check is just a copy-paste from the `if (owmlPath)` block, with very minor differences. Extract it to a separate function to avoid repeating so much code. And try to do the same for the rest of your changes if possible.

File: app/services/get-mod-database.ts [hunk 18]
```diff
@@ -52,15 +54,19 @@ export async function getModDatabase(
...
+        ? `${alphaPath}/BepInEx/plugins/${uniqueName}`
+        : `${owmlPath}/Mods/${uniqueName}`;
       const mod: Mod = {
         name,
         author,
         uniqueName,
         parent,
         remoteVersion: coerce(version)?.version ?? version,
         // TODO doesnt make sense for this to be here in remote mods
-        modPath: `${owmlPath}/Mods/${uniqueName}`,
```
Comment (Raicuparta) [inline app/services/get-mod-database.ts:63]: you moved the code without moving the comment :) Gotta be careful with that.

File: app/services/manifest.ts [hunk 19]
```diff
@@ -3,11 +3,18 @@ import { uniqueId } from 'lodash';
 export function manifestPartialToFull(partialManifest: Partial<Manifest>) {
   const missingAttributes: string[] = [];
 
-  function getAttribute(key: keyof Manifest, isUnique?: boolean): string {
+  function getAttribute(
+    key: keyof Manifest,
+    isUnique?: boolean,
+    isOptional?: boolean
```
Comment (Raicuparta) [inline app/services/manifest.ts:9]: what is this `isOptional` for? Doesn't seem to ever be used.

File: app/services/mods.ts [hunk 20]
```diff
@@ -46,14 +46,28 @@ export function cleanup(mod: Mod, tempManifestPath: string) {
...
+  if (mod.uniqueName === 'Alek.OWML') {
+    pathsToKeep = ['Mods', 'OWML.Config.json', 'OWML.Manifest.json'];
+  } else if (mod.uniqueName === 'bbepis.BepInEx') {
+    pathsToKeep = [
+      'BepInEx',
+      'OuterWilds_Alpha_1_2_Data',
+      'doorstop_config.ini',
+      'OuterWilds_Alpha_1_2.exe',
+      'winhttp.dll',
+      'BepInEx.Manifest.json',
```
Comment (Raicuparta) [inline app/services/mods.ts:60]: A bit confused about this. Seems like youre saying to just not delete anything if the mod being deleted is BepInEx. I understand this is because BepInEx is in the project root, but you essentially listed all the files here? Might as well just not run the delete function, no?

We have this `deleteFolderExcept` function because some mod updates will have files being deleted from one update to the next. So we can't just overwrite the mod files, we need to delete the old ones too. But, some mods also need to keep some user config files inside the mod dirs. So we have this exceptions list.

If a future BepInEx update drops `doorstop_config.ini` for instance, you would want it to be deleted, no? Then it shouldn't be in this list. Make sure you understand how this works, let me know if you have any questions.

File: app/store/mods-state/mod-dependencies-state.ts [hunk 21]
```diff
@@ -6,6 +6,9 @@ import { localModMap } from './mods-state';
 export const missingDependencyIdsState = selectorFamily({
   key: 'MissingDependencyIds',
   get: (mod: Mod) => ({ get }) => {
+    if (mod.isAlpha) {
+      return [];
+    }
```
Comment (Raicuparta) [inline app/store/mods-state/mod-dependencies-state.ts:11]: Can you explain this? Can alpha mods not have missing dependencies?

File: app/services/get-local-mods.ts (hunk 16 above)
Comment (MegaPiggy) [reply app/services/get-local-mods.ts:77]: It is needed for when the user doesn't have the CMOWA or BepInEx installed and runs the mod manager without internet.
If they do that without the hardcoded strings, the 2 mods don't show up in the mods tab.

File: app/store/mods-state/mod-dependencies-state.ts (hunk 21 above)
Comment (MegaPiggy) [reply app/store/mods-state/mod-dependencies-state.ts:11]: Oops, this was a mistake. I forgot to remove it.

File: app/services/mods.ts (hunk 20 above)
Comment (MegaPiggy) [reply app/services/mods.ts:60]: I changed `BepInEx` to `BepInEx/plugins/` but it deletes the mods anyways.

File: app/services/mods.ts [hunk 22]
```diff
@@ -46,14 +46,26 @@ export function cleanup(mod: Mod, tempManifestPath: string) {
...
+    pathsToKeep = ['Mods', 'OWML.Config.json', 'OWML.Manifest.json'];
+  } else if (mod.uniqueName === 'bbepis.BepInEx') {
+    pathsToKeep = [
+      'BepInEx',
+      'OuterWilds_Alpha_1_2_Data',
+      'OuterWilds_Alpha_1_2.exe',
+      'BepInEx.Manifest.json',
+    ];
+  } else if (mod.uniqueName === 'Locochoco.CMOWA') {
+    pathsToKeep = ['CMOWA.Manifest.json'];
```
[the remark this answers is not in the document]
Comment (MegaPiggy) [reply app/services/mods.ts:61]: There is no CMOWA config anymore.

**Comment by Raicuparta** [comment]
To be continued in #162
